src/hal/src/command/mod.rs

```diff
@@ -65,100 +64,131 @@ impl Level for Primary { }
+    /// Begin recording a one-shot secondary command buffer.
+    pub fn begin(
+        &mut self,
+        inheritance: CommandBufferInheritanceInfo<B>,
+    ) {
+        let flags = CommandBufferFlags::ONE_TIME_SUBMIT;
+        self.raw.begin(flags, inheritance);
+    }
+}
+
+impl<B: Backend, C> CommandBuffer<B, C, MultiShot, Secondary> {
+    /// Begin recording a multi-shot secondary command buffer.
+    pub fn begin(
+        &mut self,
+        allow_pending_resubmit: bool,
+        inheritance: CommandBufferInheritanceInfo<B>,
+    ) {
+        let flags = if allow_pending_resubmit {
+            CommandBufferFlags::SIMULTANEOUS_USE
+        } else {
+            CommandBufferFlags::empty()
+        };
+        self.raw.begin(flags, inheritance);
+    }
 }
 
-impl<'a, B: Backend, C, S: Shot, L: Level> CommandBuffer<'a, B, C, S, L> {
+impl<B: Backend, C, S: Shot, L: Level> CommandBuffer<B, C, S, L> {
     /// Create a new typed command buffer from a raw command pool.
-    pub unsafe fn new(raw: &'a mut B::CommandBuffer) -> Self {
+    pub unsafe fn new(raw: B::CommandBuffer) -> Self {
         CommandBuffer {
             raw,
             _marker: PhantomData,
         }
     }
 
-    /// Get a reference to the raw command buffer
-    pub fn as_raw(&self) -> &B::CommandBuffer {
-        &*self.raw
-    }
-
-    /// Get a mutable reference to the raw command buffer
-    pub fn as_raw_mut(&mut self) -> &mut B::CommandBuffer {
-        self.raw
-    }
-
     /// Finish recording commands to the command buffers.
     ///
-    /// The command buffer will be consumed and can't be modified further.
     /// The command pool must be reset to able to re-record commands.
-    pub fn finish(self) -> Submit<B, C, S, L> {
+    pub fn finish(&mut self) {
         self.raw.finish();
-        let raw = self.raw.clone();
-
-        ::std::mem::forget(self);
+    }
 
-        Submit::new(raw)
+    /*
+    /// Get a reference to the raw command buffer
+    pub fn as_raw(&self) -> &B::CommandBuffer {
+        &self.raw
     }
 
+    /// Get a mutable reference to the raw command buffer
+    pub fn as_raw_mut(&mut self) -> &mut B::CommandBuffer {
+        &mut self.raw
+    }*/
+
     /// Downgrade a command buffer to a lesser capability type.
     ///
     /// This is safe as a downgraded version can't be `submit`'ed
     /// since `submit` requires `self` by move.
-    pub fn downgrade<D>(&mut self) -> &mut CommandBuffer<'a, B, D, S>
+    pub fn downgrade<D>(&mut self) -> &mut CommandBuffer<B, D, S>
     where
         C: Supports<D>
     {
         unsafe { ::std::mem::transmute(self) }
     }
 }
 
-impl<'a, B: Backend, C, S: Shot> CommandBuffer<'a, B, C, S, Primary> {
+impl<B: Backend, C, S: Shot> CommandBuffer<B, C, S, Primary> {
     /// Identical to the `RawCommandBuffer` method of the same name.
-    pub fn execute_commands<I, K>(&mut self, submits: I)
+    pub fn execute_commands<'a, I, T, K, P: Shot>(&mut self, cmd_buffers: I)
```

Comment:
Is `P` used?

Reply:
nope. I wonder why Rust doesn't complain about this

Reply:
`#![deny(unused)` will make Rust complain.

Reply:
Nope. Rust warns by default, and I don't see any warnings either. This must be a bug in `rustc` then.
Anyhow, I added the deny now, it makes sense to have it in gfx-hal crate at least.

Reply:
Indeed. `unused` doesn't cover unused type parameters in functions.